FirebaseSwiftUI/FirebaseAuthSwiftUI/Sources/Services/AuthService.swift

```diff
@@ -517,4 +531,368 @@ public extension AuthService {
+          } else {
+            continuation
+              .resume(throwing: AuthServiceError
+                .multiFactorAuth("Failed to send SMS verification code to verify phone number"))
+          }
+        }
+      }
+
+    return verificationID
+  }
+
+  func completeEnrollment(session: EnrollmentSession, verificationId: String?,
+                          verificationCode: String, displayName: String) async throws {
+    // Validate session state
+    guard session.canProceed else {
+      if session.isExpired {
+        throw AuthServiceError.multiFactorAuth("Enrollment session has expired, cannot complete enrollment")
+      } else {
+        throw AuthServiceError.multiFactorAuth("Enrollment session is not in a valid state for completion")
+      }
+    }
+
+    // Validate verification code
+    guard !verificationCode.isEmpty else {
+      throw AuthServiceError.multiFactorAuth("Verification code cannot be empty")
+    }
+
+    guard let user = auth.currentUser else {
+      throw AuthServiceError.noCurrentUser
+    }
+
+    let multiFactorUser = user.multiFactor
+
+    // Create the appropriate assertion based on factor type
+    let assertion: MultiFactorAssertion
+
+    switch session.type {
+    case .sms:
+      // For SMS, we need the verification ID
+      guard let verificationId = verificationId else {
+        throw AuthServiceError
+          .multiFactorAuth("Verification ID is required for SMS enrollment")
+      }
+
+      // Create phone credential and assertion
+      let credential = PhoneAuthProvider.provider().credential(
+        withVerificationID: verificationId,
+        verificationCode: verificationCode
+      )
+      assertion = PhoneMultiFactorGenerator.assertion(with: credential)
+
+    case .totp:
+      // For TOTP, we need the secret from the session
+      guard let totpInfo = session.totpInfo else {
+        throw AuthServiceError
+          .multiFactorAuth("TOTP info is missing from enrollment session")
+      }
+
+      // Use the stored TOTP secret from the enrollment session
+      guard let secret = session._totpSecret else {
+        throw AuthServiceError
+          .multiFactorAuth("TOTP secret is missing from enrollment session")
+      }
+
+      // The concrete type is FirebaseAuth.TOTPSecret (kept as AnyObject to avoid exposing it)
+      guard let totpSecret = secret as? TOTPSecret else {
+        throw AuthServiceError
+          .multiFactorAuth("Invalid TOTP secret type in enrollment session")
+      }
+
+      assertion = TOTPMultiFactorGenerator.assertionForEnrollment(
+        with: totpSecret,
+        oneTimePassword: verificationCode
+      )
+    }
+
+    // Complete the enrollment
+    try await user.multiFactor.enroll(with: assertion, displayName: displayName)
+    currentUser = auth.currentUser
+  }
+
+  func reauthenticateCurrentUser(on user: User) async throws {
+    if let providerId = signedInCredential?.provider {
```

Comment:
we can use a guard here instead throw AuthServiceError.reauthenticationRequired("Recent login required to perform this operation.")